pkg/controllers/user/systemimage/systemimage.go

```diff
@@ -0,0 +1,187 @@
+	clusterName      string
+	daemonsets       v1beta2.DaemonSetInterface
+	daemonsetLister  v1beta2.DaemonSetLister
+	deployments      v1beta2.DeploymentInterface
+	deploymentLister v1beta2.DeploymentLister
+	projectLister    v3.ProjectLister
+	projects         v3.ProjectInterface
+}
+
+func (s *Syncer) Sync(key string, obj *v3.Project) error {
+	projects, err := s.projectLister.List(s.clusterName, systemProjectLabels.AsSelector())
+	if err != nil {
+		return fmt.Errorf("get projects failed, %v", err)
+	}
+
+	var systemProject *v3.Project
+	for _, v := range projects {
+		if v.Spec.DisplayName == "System" {
+			systemProject = v.DeepCopy()
+		}
+	}
+
+	if systemProject == nil {
+		return nil
+	}
+
+	latestSysImageVersion, err := GetSystemImageVersion()
+	if err != nil {
+		return fmt.Errorf("get latest systemImageVersion failed: %v", err)
+	}
+
+	curSysImageVersion := systemProject.Annotations[systemImageVersionAnn]
+
+	if curSysImageVersion == latestSysImageVersion {
+		return nil
+	}
+
+	nss := settings.SystemNamespaces.Default
+	namespaces := strings.Split(nss, ",")
+	if len(namespaces) == 0 {
+		return nil
+	}
+
+	for _, n := range namespaces {
+		deps, err := s.deploymentLister.List(n, labels.NewSelector())
+		if err != nil {
+			return err
+		}
+
+		daes, err := s.daemonsetLister.List(n, labels.NewSelector())
+		if err != nil {
+			return err
+		}
+
+		for _, d := range deps {
+			updated, newContainers, err := getNewContainers(d.Spec.Template.Spec.Containers)
+			if err != nil {
+				return err
+			}
+
+			if updated {
+				newdep := d.DeepCopy()
+				newdep.Spec.Template.Spec.Containers = newContainers
+				_, err = s.deployments.Update(newdep)
+				if err != nil {
+					return fmt.Errorf("upgrade deployment %s:%s failed, %v", d.Namespace, d.Name, err)
+				}
+			}
+		}
+
+		for _, d := range daes {
+			updated, newContainers, err := getNewContainers(d.Spec.Template.Spec.Containers)
+			if err != nil {
+				return err
+			}
+
+			if updated {
+				newdep := d.DeepCopy()
+				newdep.Spec.Template.Spec.Containers = newContainers
+				_, err = s.daemonsets.Update(newdep)
+				if err != nil {
+					return fmt.Errorf("upgrade daemonset %s:%s failed, %v", d.Namespace, d.Name, err)
+				}
+			}
+		}
+	}
+
+	newProject := obj.DeepCopy()
+	newProject.Annotations[systemImageVersionAnn] = latestSysImageVersion
+	_, err = s.projects.Update(newProject)
+	return err
+}
+
+func GetSystemImageVersion() (string, error) {
+	imgMap, err := getSystemImagesMap(v3.ToolsSystemImages)
+	if err != nil {
+		return "", err
+	}
+
+	b, err := json.Marshal(imgMap)
+	if err != nil {
+		return "", fmt.Errorf("marshal toolsSystemImages failed: %v", err)
+	}
+	hasher := md5.New()
+	if _, err = hasher.Write(b); err != nil {
+		return "", fmt.Errorf("write hasher failed: %v", err)
+	}
+
+	return hex.EncodeToString(hasher.Sum(nil)), nil
+}
+
+func getNewContainers(containers []corev1.Container) (updated bool, newContainers []corev1.Container, err error) {
```

Comment:
I guess we should rethink the way we do the upgrades. Ideally all the CI/CD service definitions should come as go templatized yml files, where system images are variables. Then in the go code once we figure out that the system image got changed, we just reapply the yml spec. This way it is more flexible, and can take care not only of image updates, but any other updates like labels/env vars, etc